tests/d3d12_copy.c

```diff
@@ -1684,6 +1684,229 @@ void test_multisample_resolve_formats(void)
+        ok(hr == S_OK, "Failed to create resolve destination.\n");
+
+        memset(&rtv_desc, 0, sizeof(rtv_desc));
+        rtv_desc.ViewDimension = D3D12_RTV_DIMENSION_TEXTURE2DMS;
+        rtv_desc.Format = tests[i].src_format;
+
+        ID3D12Device_CreateRenderTargetView(context.device, resolve_src, &rtv_desc, src_rtv);
+
+        rtv_desc.ViewDimension = D3D12_RTV_DIMENSION_TEXTURE2D;
+        rtv_desc.Format = tests[i].dst_format;
+
+        ID3D12Device_CreateRenderTargetView(context.device, resolve_dst, &rtv_desc, dst_rtv);
+
+        pso_desc.RTVFormats[0] = tests[i].src_format;
+
+        hr = ID3D12Device_CreateGraphicsPipelineState(context.device, &pso_desc, &IID_ID3D12PipelineState, (void**)&pso);
+        ok(hr == S_OK, "Failed to create graphics pipeline, hr %#x.\n");
+
+        ID3D12GraphicsCommandList_ClearRenderTargetView(context.list, src_rtv, green, 0, NULL);
+        ID3D12GraphicsCommandList_ClearRenderTargetView(context.list, dst_rtv, red, 0, NULL);
+
+        ID3D12GraphicsCommandList_OMSetRenderTargets(context.list, 1, &src_rtv, TRUE, NULL);
+        ID3D12GraphicsCommandList_SetGraphicsRootSignature(context.list, context.root_signature);
+        ID3D12GraphicsCommandList_SetPipelineState(context.list, pso);
+        ID3D12GraphicsCommandList_IASetPrimitiveTopology(context.list, D3D_PRIMITIVE_TOPOLOGY_TRIANGLELIST);
+        ID3D12GraphicsCommandList_RSSetViewports(context.list, 1, &viewport);
+        ID3D12GraphicsCommandList_RSSetScissorRects(context.list, 1, &scissor);
+        ID3D12GraphicsCommandList_DrawInstanced(context.list, 3, 1, 0, 0);
+
+        transition_resource_state(context.list, resolve_dst, D3D12_RESOURCE_STATE_RENDER_TARGET, D3D12_RESOURCE_STATE_RESOLVE_DEST);
+        transition_resource_state(context.list, resolve_src, D3D12_RESOURCE_STATE_RENDER_TARGET, D3D12_RESOURCE_STATE_RESOLVE_SOURCE);
+
+        ID3D12GraphicsCommandList_ResolveSubresource(context.list, resolve_dst, 0,
+                resolve_src, 0, tests[i].resolve_format);
+
+        transition_resource_state(context.list, resolve_dst, D3D12_RESOURCE_STATE_RESOLVE_DEST, D3D12_RESOURCE_STATE_COPY_SOURCE);
+
+        check_sub_resource_uint(resolve_dst, 0, context.queue, context.list, tests[i].expected, 1);
+        reset_command_list(context.list, context.allocator);
+
+        ID3D12PipelineState_Release(pso);
+
+        ID3D12Resource_Release(resolve_dst);
+        ID3D12Resource_Release(resolve_src);
+    }
+
+    for (i = 0; i < ARRAY_SIZE(invalid_tests); i++)
+    {
+        vkd3d_test_set_context("Test %u", i);
+
+        memset(&resource_desc, 0, sizeof(resource_desc));
+        resource_desc.Dimension = D3D12_RESOURCE_DIMENSION_TEXTURE2D;
+        resource_desc.Width = 4;
+        resource_desc.Height = 4;
+        resource_desc.DepthOrArraySize = 1;
+        resource_desc.MipLevels = 1;
+        resource_desc.SampleDesc.Count = 2;
```

Comment:
is MSAA 2x guaranteed to be supported? Would also work to use 4x MSAA and write sample_index / 2u in shader I think.

Reply:
*technically* it's not required, but I also can't think of a single implementation that doesn't support 2x here. Can change it though.

Reply:
Should be trivial to use 4x here, so I'd rather we do that.

Reply:
ok, updated.